File: packages/python/pomlight/style.py
from __future__ import annotations

import json
from .xml_parser import XmlElement
from .types import State, StyleSheet
from .tags import normalize_tag


def pre_parse_stylesheet(node: XmlElement, state: State) -> State:
    for child in node.children:
        if isinstance(child, str):
            continue
        if normalize_tag(child.tag) == "stylesheet":
            text = "".join(c for c in child.children if isinstance(c, str)).strip()
            if not text:
                continue
            try:
                parsed: StyleSheet = json.loads(text)
                new_styles = {**(state.styles or {}), **parsed}
                return State(
                    ctx=state.ctx, depth=state.depth, file_path=state.file_path,
                    chat=state.chat, presentation=state.presentation,
                    serializer=state.serializer, styles=new_styles,
                    sideband=state.sideband,
                )
            except (json.JSONDecodeError, TypeError):
                pass
    return state


def get_style_prop(selector: str, prop: str, state: State) -> str | None:
    if not state.styles:
        return None
    entry = state.styles.get(selector)
    if not entry:
        return None
    return entry.get(prop)


def apply_stylesheet(node: XmlElement, state: State) -> XmlElement:
    if not state.styles:
        return node
    tag = normalize_tag(node.tag)
    class_name = node.attrs.get("className") or node.attrs.get("class-name")

    merged: dict[str, str] = {}
    tag_entry = state.styles.get(tag)
    if tag_entry:
        merged.update(tag_entry)
    if class_name:
        class_entry = state.styles.get(f".{class_name}")
        if class_entry:
            merged.update(class_entry)

    if not merged:
        return node

    # Explicit attrs take precedence
    new_attrs = {**merged, **node.attrs}
    return XmlElement(tag=node.tag, attrs=new_attrs, children=node.children)


def apply_text_transform(text: str, transform: str | None = None) -> str:
    if transform == "upper":
        return text.upper()
    if transform == "lower":
        return text.lower()
    if transform == "capitalize":
        return text[0].upper() + text[1:] if text else text
    return text

File: packages/deno/src/style.ts
import type { XmlElement } from "./xml.ts";
import type { State, StyleSheet } from "./types.ts";
import { normalizeTag } from "./tags.ts";

// ---------------------------------------------------------------------------
// Stylesheet helpers
// ---------------------------------------------------------------------------

export function preParseStylesheet(node: XmlElement, state: State): State {
  for (const child of node.children) {
    if (typeof child !== "string" && normalizeTag(child.tag) === "stylesheet") {
      const text = child.children.filter((c): c is string => typeof c === "string").join("").trim();
      if (!text) continue;
      try {
        const parsed = JSON.parse(text) as StyleSheet;
        return { ...state, styles: { ...(state.styles ?? {}), ...parsed } };
      } catch { /* ignore parse errors */ }
    }
  }
  return state;
}

export function getStyleProp(selector: string, prop: string, state: State): string | undefined {
  if (!state.styles) return undefined;
  const entry = state.styles[selector];
  if (!entry) return undefined;
  return entry[prop];
}

/**
 * Merge stylesheet attributes into an element's attrs.
 * Stylesheet selectors match by tag name or by className (`.className`).
 * Explicit node attributes take precedence over stylesheet values.
 */
export function applyStylesheet(node: XmlElement, state: State): XmlElement {
  if (!state.styles) return node;
  const tag = normalizeTag(node.tag);
  const className = node.attrs.className ?? node.attrs["class-name"];

  // Collect matching stylesheet entries (tag-based, then className-based)
  const merged: Record<string, string> = {};
  const tagEntry = state.styles[tag];
  if (tagEntry) Object.assign(merged, tagEntry);
  if (className) {
    const classEntry = state.styles[`.${className}`];
    if (classEntry) Object.assign(merged, classEntry);
  }

  if (Object.keys(merged).length === 0) return node;

  // Explicit attrs take precedence
  const newAttrs = { ...merged, ...node.attrs };
  return { ...node, attrs: newAttrs };
}

export function applyTextTransform(text: string, transform?: string): string {
  switch (transform) {
    case "upper": return text.toUpperCase();
    case "lower": return text.toLowerCase();
    case "capitalize": return text.charAt(0).toUpperCase() + text.slice(1);
    default: return text;
  }
}
